src/rust/engine/fs/store/src/lib.rs

```diff
@@ -791,19 +791,26 @@ impl Store {
       .load_file_bytes_with(
         digest,
         move |bytes| {
-          OpenOptions::new()
-            .create(true)
-            .write(true)
-            .mode(if is_executable { 0o755 } else { 0o644 })
-            .open(&destination)
-            .and_then(|mut f| {
-              f.write_all(&bytes)?;
-              // See `materialize_directory`, but we fundamentally materialize files for other
-              // processes to read; as such, we must ensure data is flushed to disk and visible
-              // to them as opposed to just our process.
-              f.sync_all()
-            })
-            .map_err(|e| format!("Error writing file {:?}: {:?}", destination, e))
+          if destination.exists() {
+            std::fs::remove_file(&destination)
```

Comment:
This technique just narrows the window of conflict punting the problem down the road a bit. What do you think about writing to a temp file and using an atomic rename?: https://doc.rust-lang.org/std/fs/fn.rename.html

Reply:
Ah yeah that seems like a better idea.

Reply:
I've started to go down a bit of  rabbit hole on implementing atomic rename - apparently on POSIX it's not actually possible to atomically rename a file and also atomically ensure that the file has the permissions we set (cf. https://stackoverflow.com/questions/49031335/how-to-preserve-ownership-and-permissions-when-doing-an-atomic-file-replace ). I suppose we could check if a file with the `destination` path already exists, set the right permissions on it if so, and then create a temp file and atomically-rename, but that adds some complexity to the code without getting us anything I think.

Reply:
SGTM, this is probably not worth it right now. You've covered the nominal case of over-writing a v1 product with a v2 product and the racey case left is not one known to be exercised.